x-pack/plugins/maps/public/classes/sources/mvt_single_layer_vector_source/mvt_field_config_editor.tsx

```diff
@@ -0,0 +1,225 @@
+  state: State = {
+    currentFields: [],
+    previousFields: [],
+  };
+
+  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
+    if (_.isEqual(nextProps.fields, prevState.previousFields)) {
+      return null;
+    }
+    const clonedFields = _.cloneDeep(nextProps.fields);
+    return {
+      currentFields: clonedFields,
+      previousFields: clonedFields,
+    };
+  }
+
+  _notifyChange = _.debounce(() => {
+    const invalid = this.state.currentFields.some((field: MVTFieldDescriptor) => {
+      return field.name === '';
+    });
+
+    if (!invalid) {
+      this.props.onChange(this.state.currentFields);
+    }
+  });
+
+  _fieldChange(newFields: MVTFieldDescriptor[]) {
+    this.setState(
+      {
+        currentFields: newFields,
+      },
+      this._notifyChange
+    );
+  }
+
+  _removeField(index: number) {
+    const newFields: MVTFieldDescriptor[] = this.state.currentFields.slice();
+    newFields.splice(index, 1);
+    this._fieldChange(newFields);
+  }
+
+  _addField = () => {
+    const newFields: MVTFieldDescriptor[] = this.state.currentFields.slice();
+    newFields.push({
+      type: MVT_FIELD_TYPE.STRING,
+      name: '',
+    });
+    this._fieldChange(newFields);
+  };
+
+  _renderFieldTypeDropDown(mvtFieldConfig: MVTFieldDescriptor, index: number) {
+    const onChange = (type: MVT_FIELD_TYPE) => {
+      const newFields = this.state.currentFields.slice();
+      newFields[index] = {
+        type,
+        name: newFields[index].name,
+      };
+      this._fieldChange(newFields);
+    };
+
+    const addButton = (
+      <EuiButtonIcon
+        iconType="trash"
+        color="danger"
+        onClick={() => {
+          this._removeField(index);
+        }}
+        title={i18n.translate('xpack.maps.mvtSource.trashButtonTitle', {
+          defaultMessage: 'Remove field',
+        })}
+        aria-label={i18n.translate('xpack.maps.mvtSource.trashButtonAriaLabel', {
+          defaultMessage: 'Remove field',
+        })}
+      />
+    );
+    return (
+      <EuiSuperSelect
+        options={FIELD_TYPE_OPTIONS}
+        valueOfSelected={mvtFieldConfig.type}
+        onChange={(value) => onChange(value)}
+        append={addButton}
+      />
+    );
+  }
+
+  _renderFieldNameInput(mvtFieldConfig: MVTFieldDescriptor, index: number) {
+    const onChange = (e: ChangeEvent<HTMLInputElement>) => {
+      const name = e.target.value;
+      const newFields = this.state.currentFields.slice();
+      newFields[index] = {
+        name,
+        type: newFields[index].type,
+      };
+      this._fieldChange(newFields);
+    };
+
+    const emptyName = mvtFieldConfig.name === '';
+    let hasDupes = false;
+    for (let i = 0; i < this.state.currentFields.length; i++) {
+      if (i !== index && mvtFieldConfig.name === this.state.currentFields[i].name) {
+        hasDupes = true;
+        break;
+      }
+    }
+    const isInvalid = emptyName || hasDupes;
+    const placeholderText = isInvalid
```

Comment:
placeholder does not need this logic. Html placeholder attribute only shows when "content to be appear in the form control when the form control is empty" https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input